enhancements/builds/node-cache-for-builds.md

```diff
@@ -0,0 +1,138 @@
+---
+title: node-cache-for-builds
+authors:
+  - "@adambkaplan"
+reviewers:
+  - "@nalind
+  - "@rhdan"
+  - "@TomSweeneyRedHat"
+  - "@bparees"
+approvers:
+  - "@bparees"
+creation-date: 2019-02-18
+last-updated: 2019-02-18
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Node Cache for Builds
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. Can cri-o be enhanced so pods are not scheduled if the kubelet/node garbage collection is in progress?
+
+## Summary
+
+Builds in OpenShift 4.x currently require every layer of the base image to be pulled from its
+source registry. In OCP 3.x builds could take advantage of image layers that were stored on the
+build's node. This proposal provides a means of restoring this node-level cache.
+
+## Motivation
+
+### Goals
+
+1. Provide a means for builds to take advantage of image layers that are stored on the node.
+
+### Non-Goals
+
+1. Provide a means for builds to share an image layer cache within a namespace.
+2. Provide a means for the internal registry to pre-populate an image layer cache within a
+   namespace or across a cluster.
+3. Add persistent volume support for builds.
+
+## Proposal
+
+Build pods will mount the image layer cache that exists on every OpenShift node into the build pod.
+Buildah - running within the build pod - will use this layer cache to reduce the number of layers
+that need to be pulled when the build commences.
+
+### User Story
+
+As a developer using OpenShift to build container images
+I want builds to use image layers cached on the node
+So that builds don't have to pull all layers in their base images
+  And builds run faster.
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. The build controller will mount the node image cache located at
+   `/var/lib/containers/storage` into the build pod as `readOnly`. The mount point should not
+   conflict with buildah's own default storage directory. Ex: `/var/lib/shared/containers/storage`.
+2. The openshift/builder image is modified to configure additional image stores for buildah (via
+   `/etc/containers/storage.conf`).
+3. When using the additional image store, buildah _copies_ any blobs found in the cache to its own
+   image store within the container.
+
+### Risks and Mitigations
```

Comment:
additional risk:  unbound image cache growth.  mitigation: node cache is read-only(so builds don't contribute to it, and GC continues to manage it), builds make a temporary copy of the cached content they need, temporary copy is deleted at the end of the build.

Reply:
is there any issue ensuring that the build pod has access to the layer?

Reply:
the layer that exists on the node?  we're hostpath mounting it, i guess we'll need to make sure all the permissions are right but there shouldn't be a technical obstacle.  (whether it means giving the build pod more permissions than we'd like to is another question, though)

Reply:
@derekwaynecarr @bparees is this something that we address separately in a proof of concept or during implementation?

Reply:
it's not clear to me that there's anything to implement for either of the issues raised.  So i would just expect the EP to list the issue, describe why it won't be a problem given our architecture, and move on.

Reply:
Added mitigations below:

1. Host path doesn't exist (create it)
2. Host path does not allow privileged containers to read (worst case fall back to current behavior)